### PR TITLE
[CALCITE-7109] Support bitwise leftshifit(<<) operator And Implement LEFT_SHIFT function in SQL

### PR DESCRIPTION
**Pull request by Dwrite**
Adds support for the bitwise left shift (<<) operator in SQL expressions (PostgreSQL-compatible).

This change introduces full PostgreSQL-style semantics for LEFTSHIFT and its symbolic alias << in Apache Calcite SQL expressions.

Key changes
New operator: Introduce SHIFT_LEFT as a symbolic alias for the bitwise left shift (<<) operation.

Parser update: Extend the SQL parser to recognize << with proper precedence (32), matching PostgreSQL operator precedence rules.

Operator registration: Add SHIFT_LEFT to SqlStdOperatorTable with type inference and coercion rules:

INTEGER << INTEGER → BIGINT (ensures result fits in wider type)

BINARY << INTEGER → BINARY (preserves original precision and length)

Function implementation: Add SqlFunctions.leftShift() for:

Primitive integer types (PostgreSQL semantics: shift count must be non-negative)

Binary values (ByteString), treating them as a single big-endian bit array while preserving length

Overflow and shift count handling:

Match PostgreSQL semantics:

Negative shift count → error

Shift count ≥ bit width → result is zero for integers

No masking of shift counts (unlike Java’s <<), to stay consistent with PostgreSQL behavior

Unit tests: Add comprehensive tests covering:

Basic functionality and type coercion

Large shift counts and zero/negative shift count errors

Multi-byte BINARY values (PostgreSQL BYTEA-style)

Null propagation behavior

Edge cases (shift count exactly equal to bit width, max/min integer values)

Note: The BINARY (BYTEA) implementation follows PostgreSQL bit-string shifting behavior, operating on the entire byte array as a single unit while preserving its length.

Fixes CALCITE-7109.

### CODE REVIEW

File: core/src/test/resources/sql/operator.iq [hunk 1]
```diff
@@ -702,4 +732,57 @@ FOO BOOLEAN(1) NOT NULL
 
 !ok
 
+-- Bitwise LEFT SHIFT operator `<<`
+
+SELECT 1 << 0 AS foo;
++-----+
```
Comment (xiedeyantu) [inline core/src/test/resources/sql/operator.iq]: You can write these in one SQL.

Comment (Dwrite) [reply core/src/test/resources/sql/operator.iq]: sure.updated

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java [hunk 2]
```diff
@@ -739,6 +741,8 @@ void populate1() {
           NullPolicy.STRICT);
       defineMethod(BITXOR, BuiltInMethod.BIT_XOR.method,
           NullPolicy.STRICT);
+      defineMethod(LEFTSHIFT, LEFT_SHIFT.method,
+          NullPolicy.STRICT);
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java]: Right between two methods that start with BIT??! really? 

Use the BuilInMethod qualifier, like the other code.

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java [hunk 3]
```diff
@@ -457,6 +457,7 @@
 import static org.apache.calcite.sql.fun.SqlStdOperatorTable.LAST_DAY;
 import static org.apache.calcite.sql.fun.SqlStdOperatorTable.LAST_VALUE;
 import static org.apache.calcite.sql.fun.SqlStdOperatorTable.LEAD;
+import static org.apache.calcite.sql.fun.SqlStdOperatorTable.LEFTSHIFT;
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:462]: SHIFT_LEFT would be a better name, throughout. (Someday we’ll have SHIFT_RIGHT, possibly SHIFT_RIGHT_LOGICAL etc.)

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java [hunk 4]
```diff
@@ -541,6 +542,7 @@
 import static org.apache.calcite.sql.fun.SqlStdOperatorTable.UPPER;
 import static org.apache.calcite.sql.fun.SqlStdOperatorTable.USER;
 import static org.apache.calcite.sql.fun.SqlStdOperatorTable.VARIANTNULL;
+import static org.apache.calcite.util.BuiltInMethod.LEFT_SHIFT;
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java]: Remove this import.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 5]
```diff
@@ -3416,6 +3417,111 @@ public static ByteString bitNot(ByteString b) {
 
     return new ByteString(result);
   }
+  /**
+   * Checks if the shift count is negative.
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: It doesn’t “check”, it throws. 

There needs to be a blank line before this method. You have messed up blank lines in a couple other places.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 6]
```diff
@@ -16443,6 +16443,104 @@ private static void checkLogicalOrFunc(SqlOperatorFixture f) {
...
+    f.checkFails("16 << -2", "Shift count < 0: -2", true);
+
+    // Zero shift
+    f.checkScalar("0 << 32", "0", "INTEGER NOT NULL");
+    f.checkScalar("0 << 100", "0", "INTEGER NOT NULL");
+
+    // Max before overflow
+    f.checkScalar("CAST(63 AS TINYINT) << CAST(1 AS TINYINT)", "126", "TINYINT NOT NULL");
+    f.checkScalar("CAST(16383 AS SMALLINT) << CAST(1 AS SMALLINT)", "32766", "SMALLINT NOT NULL");
+    f.checkScalar("1073741823 << 1", "2147483646", "INTEGER NOT NULL");
```
Comment (julianhyde) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java]: I’d like to see tests where BIGINT overflows. 

Also tests where INTEGER is shifted 31 places, 32, 33 places. Overflow is difficult to detect. Similarly SMSLLINT, TINYINT. 

Is shift 0 legal? You don’t test it.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java [hunk 7]
```diff
@@ -1298,7 +1298,6 @@ public class SqlStdOperatorTable extends ReflectiveSqlOperatorTable {
       SqlBasicFunction.create("BITXOR", SqlKind.BITXOR,
           ReturnTypes.LARGEST_INT_OR_FIRST_NON_NULL,
           OperandTypes.INTEGER_INTEGER.or(OperandTypes.BINARY_BINARY));
-
   /**
```
Comment (xiedeyantu) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java]: Be careful not to delete blank lines at will.

File: core/src/test/resources/sql/operator.iq [hunk 8]
```diff
@@ -76,6 +76,36 @@ select * from "scott".emp where not sal > 1300 and not sal < 1200;
 
 !ok
 
+# [CALCITE-7109] Add support for << operator in Calcite
+SELECT CAST(empno AS INTEGER) << 1 AS foo FROM "scott".emp LIMIT 5;
```
Comment (xiedeyantu) [inline core/src/test/resources/sql/operator.iq]: I think these two case also can merge to one.

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java (hunk 3 above)
Comment (Dwrite) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:462]: sure.updated.

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java (hunk 4 above)
Comment (Dwrite) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java]: sure.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 6 above)
Comment (Dwrite) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java]: sure. updated

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java (hunk 5 above)
Comment (Dwrite) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: sure. updated

File: core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java (hunk 7 above)
Comment (Dwrite) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java]: sure.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 9]
```diff
@@ -16443,6 +16443,115 @@ private static void checkLogicalOrFunc(SqlOperatorFixture f) {
     f.checkAgg("logical_or(x)", values4, isNullValue());
   }
 
+  @Test void testLeftShiftScalarFunc() {
```
Comment (caicancai) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:16527]: Add jira link

Comment (Dwrite) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:16527]: sure.updated

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java [hunk 10]
```diff
@@ -895,6 +897,9 @@ void populate1() {
       defineMethod(TYPEOF, BuiltInMethod.TYPEOF.method, NullPolicy.STRICT);
       defineMethod(VARIANTNULL, BuiltInMethod.VARIANTNULL.method, NullPolicy.STRICT);
 
+      // shift
+      defineMethod(LEFTSHIFT, BuiltInMethod.LEFT_SHIFT.method, NullPolicy.STRICT);
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:905]: Can you add a comment describing the difference between these two?

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 11]
```diff
@@ -3506,6 +3506,152 @@ private static ByteString binaryOperator(
     return new ByteString(result);
   }
 
+  /**
+   * Performs bitwise shift on two integers.
+   * If the shift amount is positive, a left shift is applied.
+   * If the shift amount is negative, a logical right shift is applied.
+   * The shift amount is taken modulo 32.
+   */
+
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: not crazy about the extra blank line

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 12]
```diff
@@ -3506,6 +3506,152 @@ private static ByteString binaryOperator(
...
+        for (int i = bytes.length - 1 - byteShift; i >= 0; i--) {
+          int current = (bytes[i] & 0xFF) << bitShift;
+          result[i + byteShift] = (byte) ((current | carry) & 0xFF);
+          carry = (current >> 8) & 0xFF;
+        }
+        if (byteShift > 0) {
+          result[byteShift - 1] = (byte) carry;
+        }
+      }
+
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:3585]: I would remove this empty line too

File: site/_docs/reference.md [hunk 13]
```diff
@@ -2828,6 +2828,7 @@ In the following:
 | * | BITAND(value1, value2)                         | Returns the bitwise AND of *value1* and *value2*. *value1* and *value2* must both be integer or binary values. Binary values must be of the same length.
 | * | BITOR(value1, value2)                          | Returns the bitwise OR of *value1* and *value2*. *value1* and *value2* must both be integer or binary values. Binary values must be of the same length.
 | * | BITXOR(value1, value2)                         | Returns the bitwise XOR of *value1* and *value2*. *value1* and *value2* must both be integer or binary values. Binary values must be of the same length.
+| * | LEFTSHIFT(value1, value2)                      | Returns the result of left-shifting *value1* by *value2* bits. *value1* can be integer, unsigned integer, or binary. For binary, the result has the same length as *value1*. |
```
Comment (mihaibudiu) [inline site/_docs/reference.md]: please describe what happens for negative shift amounts.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 14]
```diff
@@ -3506,6 +3506,152 @@ private static ByteString binaryOperator(
...
+    byte[] result = leftShift(bytes.getBytes(), y);
+    return new ByteString(result);
+  }
+
+  /**
+   * Performs bitwise shift on a byte array.
+   * A positive shift amount performs a left shift.
+   * A negative shift amount performs a logical right shift.
+   */
+  public static byte[] leftShift(byte[] bytes, int y) {
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:3556]: this function is quite involved, can you please add a few unit tests just for it.
The unit tests could have a for loop to iterate over a set of values of y.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 15]
```diff
@@ -16518,6 +16518,218 @@ private static void checkLogicalOrFunc(SqlOperatorFixture f) {
...
+    f.checkScalar("CAST(-5 AS TINYINT) << CAST(2 AS TINYINT)", "-20", "TINYINT NOT NULL");
+
+    // === Verify return type matches first argument type ===
+    f.checkType("CAST(2 AS TINYINT) << CAST(3 AS TINYINT)", "TINYINT NOT NULL");
+    f.checkType("CAST(2 AS SMALLINT) << CAST(3 AS SMALLINT)", "SMALLINT NOT NULL");
+    f.checkType("CAST(2 AS INTEGER) << CAST(3 AS INTEGER)", "INTEGER NOT NULL");
+    f.checkType("CAST(2 AS BIGINT) << CAST(3 AS BIGINT)", "BIGINT NOT NULL");
+
+    // === BigInt shifts with explicit BIGINT inputs ===
+    f.checkScalar("CAST(1 AS BIGINT) << 62", "4611686018427387904", "BIGINT NOT NULL"); // 2^62
```
Comment (mihaibudiu) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java]: how were these results validated?

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 16]
```diff
@@ -16518,6 +16518,218 @@ private static void checkLogicalOrFunc(SqlOperatorFixture f) {
...
+    f.checkScalar("CAST(1 AS INTEGER UNSIGNED) << -1",
+        "0", "INTEGER UNSIGNED NOT NULL");
+
+
+    // === Negative shift counts ===
+    f.checkScalar("8 << -1", "4", "INTEGER NOT NULL");
+    f.checkScalar("16 << -2", "4", "INTEGER NOT NULL");
+
+    // === Shift by zero and large (but valid) shifts ===
+    f.checkScalar("0 << 32", "0", "INTEGER NOT NULL");
```
Comment (mihaibudiu) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:16600]: can you also shift a non-zero value with a large amount?

**Review by mihaibudiu — APPROVED** [review]
This is pretty good, I only left some minor comments.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 15 above)
Comment (Dwrite) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java]: The expected value "4611686018427387904" is the mathematically computed result of 2^62, 
used here to validate that the LEFT_SHIFT implementation produces the correct output 
during unit tests/build-time verification.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 16 above)
Comment (Dwrite) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:16600]: Added

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 15 above)
Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java]: Well, If you were computing 2^62 in Java it would be easier to check.
Maybe that's the right way, using the BigInteger library.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java (hunk 14 above)
Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:3556]: are you planning to do this? 
probably using the BigInteger library is also the best way to test this function.

Comment (Dwrite) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:3556]: yeah, sorry, I got little busy these days. I will fix it in the weekend

File: site/_docs/reference.md (hunk 13 above)
Comment (Dwrite) [reply site/_docs/reference.md]: Added

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java (hunk 14 above)
Comment (Dwrite) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:3556]: I think 
```
 f.checkScalar("CAST(X'4000' AS BINARY(2)) << 1", "8000", "BINARY(2) NOT NULL");
    f.checkScalar("CAST(X'0F' AS BINARY(1)) << -4", "f0", "BINARY(1) NOT NULL");
``` 
already cover this fun. Did I misunderstand?

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 15 above)
Comment (Dwrite) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java]: sure. Updated

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java (hunk 14 above)
Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:3556]: this is not a unit test for leftShift.
I meant a test which just calls this function with a lot of different input values for y and checks the result.

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java [hunk 17]
```diff
@@ -895,6 +897,17 @@ void populate1() {
       defineMethod(TYPEOF, BuiltInMethod.TYPEOF.method, NullPolicy.STRICT);
       defineMethod(VARIANTNULL, BuiltInMethod.VARIANTNULL.method, NullPolicy.STRICT);
 
+      // Left shift operations: shift bits to the left by specified amount.
+      // Supports integer, unsigned integer, and binary data types.
+      // Shift amount is normalized using modulo arithmetic based on data type bit width.
+      //
+      // "<<" operator: Infix syntax for bitwise left shift in SQL expressions (e.g., x << y)
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java]: if you will add a unit test, maybe you can split this comments into 2 parts, where each function is documented separately

**Comment by xiedeyantu** [comment]
I recommend not force-pushing a PR if it's not ready for squashing. I see @mihaibudiu has suggested some changes, but forcing a push would make it harder for him to spot the changes, and he'd have to spend more time reviewing the PR. I encourage you to keep the commits if possible. By the way, your JIRA and coding style are excellent!

**Comment by mihaibudiu** [comment]
There were only minor changes requested in the last review.

**Comment by mihaibudiu** [comment]
well, I guess the unit tests are not that minor

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java (hunk 14 above)
Comment (Dwrite) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:3556]: `f.checkScalar("LEFTSHIFT(CAST(X'ABCD' AS BINARY(2)), 0)", "abcd", "BINARY(2) NOT NULL");
    f.checkScalar("LEFTSHIFT(CAST(X'123456' AS BINARY(3)), 4)", "204163", "BINARY(3) NOT NULL");
    f.checkScalar("LEFTSHIFT(CAST(X'8000' AS BINARY(2)), 1)", "0001", "BINARY(2) NOT NULL");
    f.checkScalar("LEFTSHIFT(CAST(X'4000' AS BINARY(2)), 1)", "8000", "BINARY(2) NOT NULL");
    f.checkScalar("LEFTSHIFT(CAST(X'0F' AS BINARY(1)), -4)", "f0", "BINARY(1) NOT NULL");
`
are these cover this func ?

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:3556]: I wanted just a unit test for that function, not one that goes through all the other layers.
The unit test fills a byte array with some pattern, shifts it, and compares the result to the ground truth computed using some other means, e.g., BitInteger.

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:3556]: it shifts it with all amounts possible, from -length-2 to length+2, where length is the byte array size in bits

Comment (Dwrite) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:3556]: Thanks for clarifying earlier. I’ve added a dedicated unit test for SqlFunctions.leftShift and placed it in SqlFunctionsTest.java, following the existing Calcite test style. Could you confirm if this is the location and style you had in mind?

**Comment by mihaibudiu** [comment]
I think you can squash the commits for merging

**Comment by Dwrite** [comment]
Thanks you  guys.
